polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -1427,36 +1427,52 @@ ValueCategory MLIRScanner::CommonFieldLookup(clang::QualType CT,
-    else if (ElemTy.isa<sycl::NdRangeType>())
-      Result = SYCLCommonFieldLookup<sycl::NdRangeType>(Val, FNum, Shape);
-    else if (ElemTy.isa<sycl::ItemType>())
-      Result = SYCLCommonFieldLookup<sycl::ItemType>(Val, FNum, Shape);
-    else if (ElemTy.isa<sycl::ItemBaseType>())
-      Result = SYCLCommonFieldLookup<sycl::ItemBaseType>(Val, FNum, Shape);
-    else if (ElemTy.isa<sycl::NdItemType>())
-      Result = SYCLCommonFieldLookup<sycl::NdItemType>(Val, FNum, Shape);
-    else if (ElemTy.isa<sycl::GroupType>())
-      Result = SYCLCommonFieldLookup<sycl::GroupType>(Val, FNum, Shape);
-    else if (ElemTy.isa<sycl::AtomicType>())
-      Result = SYCLCommonFieldLookup<sycl::AtomicType>(Val, FNum, Shape);
-    else
-      llvm_unreachable("not implemented");
+    Result =
+        TypeSwitch<Type, Value>(ElemTy)
+            .Case<sycl::ArrayType>([&](sycl::ArrayType AT) {
+              assert(FNum < AT.getBody().size() && "ERROR");
+              const auto ElemType = AT.getBody()[FNum].cast<MemRefType>();
+              const auto ResultType = MemRefType::get(
+                  ElemType.getShape(), ElemType.getElementType(),
+                  MemRefLayoutAttrInterface(), MT.getMemorySpace());
+              return Builder.create<polygeist::SubIndexOp>(
+                  Loc, ResultType, Val, getConstantIndex(FNum));
+            })
+            .Case<sycl::AccessorType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::AccessorType>(Val, FNum,
+                                                               Shape);
+            })
+            .Case<sycl::AccessorImplDeviceType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::AccessorImplDeviceType>(
+                  Val, FNum, Shape);
+            })
+            .Case<sycl::AccessorSubscriptType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::AccessorSubscriptType>(
+                  Val, FNum, Shape);
+            })
+            .Case<sycl::AtomicType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::AtomicType>(Val, FNum, Shape);
+            })
+            .Case<sycl::GroupType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::GroupType>(Val, FNum, Shape);
+            })
+            .Case<sycl::ItemBaseType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::ItemBaseType>(Val, FNum,
+                                                               Shape);
+            })
+            .Case<sycl::ItemType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::ItemType>(Val, FNum, Shape);
+            })
+            .Case<sycl::NdItemType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::NdItemType>(Val, FNum, Shape);
+            })
+            .Case<sycl::NdRangeType>([&](auto ElemTy) {
+              return SYCLCommonFieldLookup<sycl::NdRangeType>(Val, FNum, Shape);
+            })
```

Comment:
```suggestion
            .Case<sycl::AccessorType, sycl::AccessorImplDeviceType,
                       sycl::AccessorSubscriptType, sycl::AtomicType,
                       sycl::GroupType, sycl::ItemBaseType,
                       sycl::ItemType, sycl::NdItemType,
                       sycl::NdRangeType>([&](auto ElemTy) {
              return SYCLCommonFieldLookup<decltype(ElemTy)>(Val, FNum, Shape);
            })
```
Sorry for being too picky, but this saves so many lines.

Reply:
That's great, I wanted to do something like that, but didn't know `decltype`, thanks for the suggestion.